### PR TITLE
qemu_cortex_m0: Timer fixes

### PR DESCRIPTION
**Pull request by carlocaione**
Please refer to the commit messages for a detailed explanation of the fixes.

Tentative fix for #57512

### CODE REVIEW

**Review by anangl — DISMISSED** [review]

**Review by gmarull — DISMISSED** [review]

**Review by microbuilder — DISMISSED** [review]

**Review by stephanosio — DISMISSED** [review]

**Comment by carlocaione** [comment]
Oh well, apparently this fix is messing up badly with a lot of other tests. Interesting.

**Comment by carlocaione** [comment]
Now, this is a pretty interesting problem. Tagging @stephanosio @microbuilder @andyross just because they are involved in a way or another.

Applying the two commits in this PR, we still have `tests/kernel/threads/no-multithreading/kernel.threads.no-multithreading` failing.

This is easily reproducible with this snippet (the `printk` is never executed and `k_busy_wait()` is never left):

```c
unsigned int key = irq_lock();
k_busy_wait(15000);
irq_unlock(key);

printk("done\n");
```
and in particular this is **only** reproducible when using `-icount`.

This works fine:
```
qemu-system-arm -cpu cortex-m0 -machine microbit -nographic -kernel build/zephyr/zephyr.elf
    -rtc clock=vm
```

This doesn't work:
```
qemu-system-arm -cpu cortex-m0 -machine microbit -nographic -kernel build/zephyr/zephyr.elf
    -rtc clock=vm -icount shift=6,align=off,sleep=off
```

Using `--trace` to see what's going on in the QEMU timer driver it is possible to see that in the non-working case the counter value is never incremented.

Any idea why the `-icount` parameter here could affect the QEMU emulation?

**Comment by pm215** [comment]
> Any idea why the -icount parameter here could affect the QEMU emulation?

In non-icount mode, the clock used by timer devices is the real-world wall clock, and guest execution is just "run as fast as possible". So a 1MHz timer will always fire after 1 microsecond real-world time (and if you program the timer to fire 'too fast' then you'll find the guest doesn't make forward progress because it's always handling timer interrupts).  In icount mode, the clock used by timer devices is tied to guest instruction execution, so (ignoring the 'auto' icount submode since you aren't using it) timer expiry deadlines (and how far a free-running clock moves forward) are calculated as "after we've executed another N guest instructions".

I don't know exactly what's going on here, but it's not super surprising that icount vs non-icount might result in you seeing different behaviour out of a timer device.

**Comment by pm215** [comment]
I'm still curious about what exactly is happening that means that your k_busy_wait() never exits when in icount mode. If you have a test binary and QEMU command line I can have a look at what's going on...

**Comment by carlocaione** [comment]
> I'm still curious about what exactly is happening that means that your k_busy_wait() never exits when in icount mode. If you have a test binary and QEMU command line I can have a look at what's going on...

Here the ELF https://drive.google.com/file/d/1EU03byaN1nPeyaBcgjWBCvenhT375rvl/view?usp=sharing

This is compiled from this PR rebased on top of `4c8b18920a7c1393c3f182af85a4232b7280f6d1` using this diff:
```diff
diff --git a/samples/hello_world/src/main.c b/samples/hello_world/src/main.c
index bc7a4c1d71..1cfcdfe456 100644
--- a/samples/hello_world/src/main.c
+++ b/samples/hello_world/src/main.c
@@ -8,6 +8,11 @@
 
 int main(void)
 {
-       printk("Hello World! %s\n", CONFIG_BOARD);
+       unsigned int key = irq_lock();
+       k_busy_wait(15000);
+       irq_unlock(key);
+
+       printk("done\n");
+
        return 0;
 }
```
With: `west build -b qemu_cortex_m0 samples/hello_world -p`

Using `QEMU emulator version 7.2.0 (Debian 1:7.2+dfsg-5ubuntu2)` I get:

- working version:
```
qemu-system-arm -cpu cortex-m0 -machine microbit -nographic -kernel build/zephyr/zephyr.elf -rtc clock=vm
```
- non-working version:
```
qemu-system-arm -cpu cortex-m0 -machine microbit -nographic -kernel build/zephyr/zephyr.elf -rtc clock=vm -icount shift=6,align=off,sleep=off
```

**Review by andyross — APPROVED** [review]
One note about CI minutiae worth checking, but everything looks fine behaviorally.

**Comment by andyross** [comment]
As far as the icount interaction, I don't know that I have much to say except that it smells like an emulation bug.  The default k_busy_wait() is just a loop around the timer driver's sys_clock_cycle_get_32() function, which for nrf_timer_timer.c is a pretty standard looking thing that just locks interrupts and calls its counter read function, which looks like this:

```c
static uint32_t counter(void) {
    nrf_timer_task_trigger(TIMER,
            nrf_timer_capture_task_get(NRF_TIMER_CC_CHANNEL1));
    return nrf_timer_cc_get(TIMER, NRF_TIMER_CC_CHANNEL1); }
}
```

The fact that it wants to "trigger" a "task" implies some level of async handling, and it seems not impossible that a tight loop might defeat that?  Maybe the emulator is waiting for the next internal state change (or whatever) to arrive before handling the trigger, which keeps being reset by the loop?

Might be worth seeing if you could reproduce the bug with only the nRF HAL calls and a bare qemu (i.e. not down in a zephyr thread in an existing test, but right after boot in or near z_cstart()).

**Review by microbuilder — APPROVED** [review]
Further investigation required, but this seems acceptable as a tentative solution. Thanks for the detailed commit messages as well.

**Comment by microbuilder** [comment]
@pm215 If you have any insights into why this is failing with `ICOUNT`, please post them here despite being merged, and a new PR or issue can be raised as appropriate. Thanks.

**Comment by pm215** [comment]
The hang with icount mode looks like a QEMU bug in the model of the nrf51 timer. Essentially it's a rounding issue -- when the guest queries the timer tick count, the code says "how many (emulated wallclock) nanoseconds since we last calculated the tick count", and converts that to ticks to see what it should add to that tick count to return. This works OK in non-icount mode since even in a tight loop querying the count, the emulated wallclock (== the real wallclock) advances by 3 or 4 ticks each time. But in icount mode the emulated wallclock is based on number of instructions executed, so the loop is so tight that it has moved on by less than a timer tick since the last iteration. So the rounding means the device model says "counter has moved forwards 0 ticks" every time... (Even in non-icount mode the rounding will be causing a loss of accuracy, but at least we make forward progress.) I'll have a look at fixing this on the QEMU end.

**Comment by pm215** [comment]
This QEMU patch should fix the icount-mode hang:
https://patchew.org/QEMU/20230606134917.3782215-1-peter.maydell@linaro.org/
It's possible it also helps with some of the other odd behaviour noted in the original bug report. If you're able to test it in your Zephyr test suite that would be very helpful.

**Comment by carlocaione** [comment]
> This QEMU patch should fix the icount-mode hang: https://patchew.org/QEMU/20230606134917.3782215-1-peter.maydell@linaro.org/ It's possible it also helps with some of the other odd behaviour noted in the original bug report. If you're able to test it in your Zephyr test suite that would be very helpful.

@stephanosio can we?

File: boards/arm/qemu_cortex_m0/qemu_cortex_m0_defconfig [hunk 1]
```diff
@@ -19,3 +19,6 @@ CONFIG_CONSOLE=y
 CONFIG_UART_CONSOLE=y
 
 CONFIG_PINCTRL=y
+
+# icount is kinda broken when the NRF timer emulation is used
+CONFIG_QEMU_ICOUNT=n
```
[the remark this answers is not in the document]
Comment (keith-packard) [reply boards/arm/qemu_cortex_m0/qemu_cortex_m0_defconfig:24]: I'm actually seeing completely consistent and reproducible failures without icount enabled. I managed to automatically bisect a failure on main back to this patch.

https://github.com/zephyrproject-rtos/zephyr/actions/runs/5914686960/job/16040717438?pr=57340

Anyone care to figure out how to solve both problems?

**Comment by microbuilder** [comment]
> > This QEMU patch should fix the icount-mode hang: https://patchew.org/QEMU/20230606134917.3782215-1-peter.maydell@linaro.org/ It's possible it also helps with some of the other odd behaviour noted in the original bug report. If you're able to test it in your Zephyr test suite that would be very helpful.
> 
> @stephanosio can we?

Just as a FYI, this has been merged into QEMU, so any update to a recent version of QEMU in the next Zephyr SDK release should include the required patch.

**Pull request merged**